README
What is it?
====================

ZoteroXmlImporter extends Zotero by providing import of data into Zotero from a
simple XML format.

"Zotero [zoh-TAIR-oh] ... a free, easy-to-use Firefox extension to help you
collect, manage, cite, and share your research sources."
(http://www.zotero.org/)

ZoteroXmlImporter is also an extension for the Firefox browser, it works as a
plugin for Zotero - you need an appropriate version of Zotero installed.


For whom is it?
====================

ZoteroXmlImporter is for you if you 

1. have data in any format that is not natively supported by Zotero, AND

2. you are able to convert your data format to the simple XML format supported
by ZoteroXmlImporter.

If you are clever enough, you better use the RDF import format of Zotero,
which is XML based, too, and has the advantage of being completely integrated
and supported by Zotero.

In the meantime, ZoteroXmlImporter may be helpful. Its elements are simply the
names of Zotero item types and fields, like

<?xml version='1.0'?>
<zotero-import>
<document>
  <title>Gardening and more</title>
  <publisher>Hans Dampf</publisher>
  <language>en-US</language>
  <tag>Top10</tag>
  <tag>Gardening</tag>
  <note>Should read this again <strong>soon!</strong></note>
</document>
</zotero-import>


Features, Restrictions
======================

- imports
  * all "regular" bibliographic items (artwork, ... webpage) with child notes
    and tags and links
  * top-level note items with tags and links
  into the main collection only (not into the currently selected sub-collection).

- does not import
  * notes for links
  * attachments
  * structured "creators" (e.g. imports one author string "Hans Dampf", no
    firstname/lastname structure)

- no checking except when each item is passed to Zotero

- rollback if import of some items fails

- no checking of field formats (date strings like "yesterday" or "meouw" are
  passed to Zotero and accepted)

- progress meter seems to come up only finally :-(

- see TODO in the code


Import format
======================

See test/generatedAll.xml for an example file.
The file has been generated by

  scripts/computeItemTypeFields.html

which also created the RelaxNG syntax description

  design/zotero-import.rnc

  NOTE that this file does not correctly describe the HTML contents of <note>
  elements (it should allow all HTML elements...).

The script also creates an HTML listing of Zotero item types and their
supported fields. You need to feed the script with a Zotero system.sql
file. The one used for the files above was from Zotero 2.0:

  https://www.zotero.org/trac/browser/extension/branches/2.0/system.sql


Installation
======================
See INSTALL


License
======================
See COPYING

--
HSteeb 2010-08-27
EOF


CHANGES
Version 0.9
(2010-08-27)

Initial version


test/computeItemTypeFields.js
function Compute( resultType)
{
  var withCreators = document.getElementById("withCreators").checked;

  var lines = document.getElementById("sql").value;
  // 207	INSERT INTO itemTypes VALUES (1,'note',NULL,0);

  var result;

  // parse itemTypes
  var itRE = /INSERT\s+INTO\s+itemTypes\s+VALUES\s*\(\s*(\d+)\s*,\s*['\"](\w+)[\"']/g;
  var itn2s = {}; // itemType number to string
  var its2n = {}; // itemType string to number
  var its = []; // itemType strings
  while ( (result = itRE.exec(lines)) != null) {
    itn2s[result[1]] = result[2];
    its2n[result[2]] = result[1];
    its.push(result[2]);
  }
  its.sort();

  // parse fields
  var fRE = /INSERT\s+INTO\s+fields\s+VALUES\s*\(\s*(\d+)\s*,\s*['\"](\w+)[\"']/g;
  var fn2s = {}; // field number to string
  while ( (result = fRE.exec(lines)) != null) {
    fn2s[result[1]] = result[2];
  }

  // parse itemTypeFields
  var itfRE = /INSERT\s+INTO\s+itemTypeFields\s+VALUES\s*\(\s*(\d+)\s*,\s*(\d+)/g;
  var itfn2N = {}; // itemType number to field numbers
  while ( (result = itfRE.exec(lines)) != null) {
    if (!itfn2N[result[1]]) {
      itfn2N[result[1]] = [];
    }
    itfn2N[result[1]].push(result[2]);
  }

  if (withCreators) {
    // parse creatorTypes
    var ctRE = /INSERT\s+INTO\s+creatorTypes\s+VALUES\s*\(\s*(\d+)\s*,\s*['\"](\w+)[\"']/g;
    var ctn2s = {}; // creatorType number to string
    var cts2n = {}; // creatorType string to number
    var cts = []; // creatorType strings
    while ( (result = ctRE.exec(lines)) != null) {
      ctn2s[result[1]] = result[2];
      cts2n[result[2]] = result[1];
      cts.push(result[2]);
    }
    cts.sort();
  
    // parse itemTypeCreatorTypes
    var itctRE = /INSERT\s+INTO\s+itemTypeCreatorTypes\s+VALUES\s*\(\s*(\d+)\s*,\s*(\d+)/g;
    var itctn2N = {}; // itemType number to creator type numbers
    while ( (result = itctRE.exec(lines)) != null) {
      if (!itctn2N[result[1]]) {
        itctn2N[result[1]] = [];
      }
      itctn2N[result[1]].push(result[2]);
    }
  }


  var s = "";
  if (resultType == "html") {
    // compute HTML
    s = "<html><body>";
  
    // table of contents
    s += "<p>\n";
    for (var i = 0; i < its.length; ++i) {
      var itemType = its[i];
      s += "<a href='#it_" + itemType + "'>" + itemType + "</a><br/>\n";
    }
    s += "</p>\n";
  
    // each itemType
    for (var j = 0; j < its.length; ++j) {
      var itemType = its[j];
      var itemTypeNumber = its2n[itemType];
      s += "<h1 id='it_" + itemType + "' name='it_" + itemType + "''>" + itemType + "</h1>\n";

      // each field
      var FieldNumbers = itfn2N[itemTypeNumber];
      if (FieldNumbers) {
        var F = [];
        // each itemTypeField (sorted)
        for (var f = 0; f < FieldNumbers.length; ++f) {
          var fieldNumber = FieldNumbers[f];
          var fieldName = fn2s[fieldNumber];
          F.push(fieldName);
        }
        F.sort();
        s += "<p>\n" + F.join("<br/>\n") + "</p>\n";
      }

      if (withCreators) {
        // each creator type
        var CreatorTypeNumbers = itctn2N[itemTypeNumber];
        if (CreatorTypeNumbers) {
          var CT = [];
          // each itemTypeCreatorType (sorted)
          for (var ct = 0; ct < CreatorTypeNumbers.length; ++ct) {
            var creatorTypeNumber = CreatorTypeNumbers[ct];
            var creatorTypeName = ctn2s[creatorTypeNumber];
            CT.push(creatorTypeName);
          }
          CT.sort();
          s += "<p>\n" + CT.join("<br/>\n") + "</p>\n";
        }
      }
    }
  }
  else if (resultType == "testfile") {
    // compute XML
    s = "<?xml version='1.0'?>\n"
      + "<zotero-import>\n";
  
    // each itemType
    for (var j = 0; j < its.length; ++j) {
      var itemType = its[j];

      // special handling for:
      // attachment
      if (itemType == "attachment") { continue; }

      var itemTypeNumber = its2n[itemType];
      s += "<" + itemType + ">\n"; // needs closing tag - do not if (!FieldNumbers) { continue; }

      // note
      if (itemType == "note") { 
        s += "<text>My example note with <strong>strong</strong> text, a list with\n<ul>\n<li>item1</li>\n<li>item2</li>\n</ul>\n</text>\n";
      }

      // each field
      var FieldNumbers = itfn2N[itemTypeNumber];
      if (FieldNumbers) {
        var F = [];
        // each itemTypeField (sorted)
        for (var f = 0; f < FieldNumbers.length; ++f) {
          var fieldNumber = FieldNumbers[f];
          var fieldName = fn2s[fieldNumber];
          F.push(fieldName);
        }
        F.sort();
        for (var k = 0; k < F.length; ++k) {
          var fieldName = F[k];
          s += "<" + fieldName + ">";
          if (fieldName == "title") {
            s += itemType + ": ";
          }
          s += fieldName + "...</" + fieldName + ">\n";
        }
      }

      if (withCreators) {
        // each creator type
        var CreatorTypeNumbers = itctn2N[itemTypeNumber];
        if (CreatorTypeNumbers) {
          var CT = [];
          // each itemTypeCreatorType (sorted)
          for (var ct = 0; ct < CreatorTypeNumbers.length; ++ct) {
            var creatorTypeNumber = CreatorTypeNumbers[ct];
            var creatorTypeName = ctn2s[creatorTypeNumber];
            CT.push(creatorTypeName);
          }
          CT.sort();
          for (var k = 0; k <CT.length; ++k) {
            var creatorTypeName = CT[k];
            s += "<" + creatorTypeName + ">" + creatorTypeName + "...</" + creatorTypeName + ">\n";
          }
        }
      }

      // add two tags as illustration
      var tagBy2 = "Tag " + (j % 2 ? "even" : "odd");
      s += "<tag>" + tagBy2 + "</tag>\n";

      var tagBy3 = (0 == j % 3) ? "A" : (1 == j % 3 ? "B" : "C");
      s += "<tag>Category " + tagBy3 + "</tag>\n";

      s += "</" + itemType + ">\n";
    } // itemTypes
    s += "</zotero-import>\n";
    s = s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/\n/g, "<br/>\n");
  }

  else if (resultType == "rnc") {
    s = ""
      + "# RelaxNG schema for the \"Zotero Xml\" import format.\n"
      + "# Helmut Steeb 2010-08-27 hs2010\@bible2.net (insert current year)\n"
      + "# Generator: computeItemTypeFields.js\n"
      + "#\n"
      + "# RESTRICTIONS:\n"
      + "# - only <note> is allowed child elements,\n"
      + "#   but the modelling is not correct, you'll get errors there.\n"
      + "# - All other terminal elements accept \"text\", no details are modelled.\n"
      + "\n\n"
      + "zotero-import = element zotero-import {\n";
    // each itemType
    for (var j = 0; j < its.length; ++j) {
      // assumes attachment is not at j==0
      if (its[j] == "attachment") { continue; }
      s += (j == 0 ? " " : "&" ) + " " + its[j] + "\n";
    }
    s += "}\n\n";

    var AllFields = [];
    var FieldsProcessed = {};
    for (var j = 0; j < its.length; ++j) {
      var itemType = its[j];

      if (itemType == "attachment") { continue; }

      s += itemType + " = element " + itemType + " {\n";
      // note
      if (itemType == "note") { 
        s += "  \\text\n"; // not rnc "text" but our element "text""
      }
      else {
        // each field
        var itemTypeNumber = its2n[itemType];
        var FieldNumbers = itfn2N[itemTypeNumber];
        if (FieldNumbers) {
          var F = [];
          // each itemTypeField (sorted)
          for (var f = 0; f < FieldNumbers.length; ++f) {
            var fieldNumber = FieldNumbers[f];
            var fieldName = fn2s[fieldNumber];
            F.push(fieldName);
          }
          F.sort();
          for (var k = 0; k < F.length; ++k) {
            var fieldName = F[k];
            s += (k == 0 ? " " : "&") + " " + fieldName + "?\n";
            if (!FieldsProcessed[fieldName]) {
              AllFields.push(fieldName);
              FieldsProcessed[fieldName] = true;
            }
          }
        }

        if (withCreators) {
          // each creator type
          var CreatorTypeNumbers = itctn2N[itemTypeNumber];
          if (CreatorTypeNumbers) {
            var CT = [];
            // each itemTypeCreatorType (sorted)
            for (var ct = 0; ct < CreatorTypeNumbers.length; ++ct) {
              var creatorTypeNumber = CreatorTypeNumbers[ct];
              var creatorTypeName = ctn2s[creatorTypeNumber];
              CT.push(creatorTypeName);
            }
            CT.sort();
            for (var k = 0; k <CT.length; ++k) {
              var creatorTypeName = CT[k];
              s += "& " + creatorTypeName + "*\n";
              if (!FieldsProcessed[creatorTypeName]) {
                AllFields.push(creatorTypeName);
                FieldsProcessed[creatorTypeName] = true;
              }
            }
          }
        } // withCreators
      } // else each field

      s += "& tag*\n";
      if (itemType != "note") {
        s += "& note*\n";
      }
      s += "& link*\n";
      s += "}\n\n";
    } // itemTypes

    s += "\n";
    s += "tag = element tag { text }\n";
    s += "\\text = element text { (text | ANY)* }\n";
    s += "ANY = element * { ANY }\n";
    s += "link = element link { text }\n";
    s += "\n";

    AllFields.sort();
    for (var f = 0, len = AllFields.length; f < len; ++f) {
      s += AllFields[f] + " = element " + AllFields[f] + " { text }\n";
    }
    s += "\nstart = zotero-import\n";
    s = "<pre>\n" + s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/\n/g, "\n") + "</pre>\n";
  }
  else {
    s = "Unknown result type";      
  }
  var Result = document.getElementById("Result");
  Result.innerHTML = s;
}  


impl/chrome/content/ZoteroXmlImporter/Dialog.js
// ===================================================================
// Mozilla Dialogs
// $Id: Dialog.js 4137 2010-06-07 19:11:56Z helmut $

var Moz;
if (!Moz) {
  Moz = {};
}
else if (typeof Moz != "object") {
  throw new Error("Moz already exists and is not an object");
}

if (!Moz.Dialog) {
  Moz.Dialog = {};
}
else if (typeof Moz.Dialog != "object") {
  throw new Error("Moz.Dialog already exists and is not an object");
}


// fileDialog:
// title = title
// parent = parent window for init() call, defaults to "window"
//   (should be set for dialog on other modal dialog
//    in order to after dialog close
//    get the focus back to the modal dialog, not to the main window)
// mode = one of Components.interfaces.nsIFilePicker.mode{Open,Save,GetFolder,OpenMultiple}
// filters = array of title/filter pairs, like
//   ["Text", "*.txt", "XML", "*.xml"],
//   note that filterAll is appended as last filter (except if suppressFilterAll is given)
// filterIndex = index initially selected filter (0-based)
// defaultString = the filename that should be suggested to the user as a default.
// displayDirectory = the directory to show initially (nsILocalFile, e.g. from Moz.FileIO.open)
// suppressFilterAll = if non-null, does not append filterAll
// wantFile = if given, returns nsIFile, otherwise (= default) returns nsIFile.path


Moz.Dialog.fileDialog = function(args)
{
  try {
    // create picker
    const nsIFilePicker = Components.interfaces.nsIFilePicker;
    var fp = Components.classes["@mozilla.org/filepicker;1"].createInstance(nsIFilePicker);

    var w = args.parent || window;
    // setup picker
    fp.init(w, args.title, args.mode);
    if (args.filters != null) {
      while (args.filters.length) {
        var title  = args.filters.shift();
        var filter = args.filters.shift();
        fp.appendFilter(title, filter);
      }
    }
    if (args.suppressFilterAll == null) {
      fp.appendFilters(nsIFilePicker.filterAll);
    }
    if (args.filterIndex != null) {
      fp.filterIndex = args.filterIndex;
    }
    if (args.defaultString != null) {
      fp.defaultString = args.defaultString;
    }
    if (args.displayDirectory) {
      fp.displayDirectory = args.displayDirectory;
    }

    // show picker
    var rv = fp.show();

    // evaluate result
    if (rv == nsIFilePicker.returnOK || rv == nsIFilePicker.returnReplace) {
      return args.wantFile ? fp.file : fp.file.path;
    }
    return null;
  }
  catch (e) {
    alert("Moz.Dialog.fileDialog: " + e.name + ": " + e.message);
    throw e;
  }
}



impl/chrome/content/ZoteroXmlImporter/FileIO.js
/////////////////////////////////////////////////
/////////////////////////////////////////////////
//
// Basic JavaScript File and Directory IO module
// By: MonkeeSage, v0.1
//
// 2008-03-01 HS: from
//   http://kb.mozillazine.org/Io.js
// 2009-03-13 HS: moved into Moz namespace, split io.js into Moz.FileIO.js and Moz.DirIO.js
/////////////////////////////////////////////////
/////////////////////////////////////////////////

/* 2009-03-13 HS: moved into Moz namespace */
var Moz;
if (!Moz) {
  Moz = {};
}
else if (typeof Moz != "object") {
  throw new Error("Moz already exists and is not an object");
}

  /////////////////////////////////////////////////
  // Basic file IO object based on Mozilla source
  // code post at forums.mozillazine.org
  /////////////////////////////////////////////////

  // Example use:
  // var fileIn = FileIO.open('/test.txt');
  // if (fileIn.exists()) {
  //   var fileOut = FileIO.open('/copy of test.txt');
  //   var str = FileIO.read(fileIn);
  //   var rv = FileIO.write(fileOut, str);
  //   alert('File write: ' + rv);
  //   rv = FileIO.write(fileOut, str, 'a');
  //   alert('File append: ' + rv);
  //   rv = FileIO.unlink(fileOut);
  //   alert('File unlink: ' + rv);
  // }

  Moz.FileIO = {

    localfileCID  : '@mozilla.org/file/local;1',
    localfileIID  : Components.interfaces.nsILocalFile,

    finstreamCID  : '@mozilla.org/network/file-input-stream;1',
    finstreamIID  : Components.interfaces.nsIFileInputStream,

    foutstreamCID : '@mozilla.org/network/file-output-stream;1',
    foutstreamIID : Components.interfaces.nsIFileOutputStream,

    sinstreamCID  : '@mozilla.org/scriptableinputstream;1',
    sinstreamIID  : Components.interfaces.nsIScriptableInputStream,

    suniconvCID   : '@mozilla.org/intl/scriptableunicodeconverter',
    suniconvIID   : Components.interfaces.nsIScriptableUnicodeConverter,

    open   : function(path) {
      try {
        var file = Components.classes[this.localfileCID]
                .createInstance(this.localfileIID);
        file.initWithPath(path);
        return file;
      }
      catch(e) {
        return false;
      }
    },

    read   : function(file, charset) {
      try {
        var data     = new String();
        var fiStream = Components.classes[this.finstreamCID]
                  .createInstance(this.finstreamIID);
        var siStream = Components.classes[this.sinstreamCID]
                  .createInstance(this.sinstreamIID);
        fiStream.init(file, 1, 0, false);
        siStream.init(fiStream);
        data += siStream.read(-1);
        siStream.close();
        fiStream.close();
        if (charset) {
          data = this.toUnicode(charset, data);
        }
        return data;
      }
      catch(e) {
        return false;
      }
    },

    write  : function(file, data, mode, charset) {
      try {
        var foStream = Components.classes[this.foutstreamCID]
                  .createInstance(this.foutstreamIID);
        if (charset) {
          data = this.fromUnicode(charset, data);
        }
        var flags = 0x02 | 0x08 | 0x20; // wronly | create | truncate
        if (mode == 'a') {
          flags = 0x02 | 0x10; // wronly | append
        }
        foStream.init(file, flags, 0664, 0);
        foStream.write(data, data.length);
        // foStream.flush();
        foStream.close();
        return true;
      }
      catch(e) {
//TODO low 2008-08-14 HS
//alert("FileIO.write: " + e);
        return false;
      }
    },

    create : function(file) {
      try {
        file.create(0x00, 0664);
        return true;
      }
      catch(e) {
        return false;
      }
    },

    unlink : function(file) {
      try {
        file.remove(false);
        return true;
      }
      catch(e) {
        return false;
      }
    },

    path   : function(file) {
      try {
        return 'file:///' + file.path.replace(/\\/g, '\/')
              .replace(/^\s*\/?/, '').replace(/\ /g, '%20');
      }
      catch(e) {
        return false;
      }
    },

    toUnicode   : function(charset, data) {
      try{
        var uniConv = Components.classes[this.suniconvCID]
                  .createInstance(this.suniconvIID);
        uniConv.charset = charset;
        data = uniConv.ConvertToUnicode(data);
      }
      catch(e) {
        // foobar!
      }
      return data;
    },

    fromUnicode : function(charset, data) {
      try {
        var uniConv = Components.classes[this.suniconvCID]
                  .createInstance(this.suniconvIID);
        uniConv.charset = charset;
        data = uniConv.ConvertFromUnicode(data);
        // data += uniConv.Finish();
      }
      catch(e) {
        // foobar!
      }
      return data;
    },

    // 2009-03-13 HS added:
    exists: function(path) // HS
    {
      var file = this.open(path);n
      return file && file.exists();
    },

    loadText: function(url) // HS
    {
      var filename = url.replace(/file:\/\//, "");
      var file = this.open(filename);
//TODO low 2008-12-23 HS - need "utf-8"?
      return this.read(file, "utf-8");
    },

    saveText: function(text, url) // HS
    {
      var filename = url.replace(/file:\/\//, "");
      var file = this.open(filename);
      // 2008-05-03 HS:
      // without passing "utf-8", the utf-8 string is saved as latin1 :-((
      if (!this.write(file, text, "", "utf-8")) {
        //alert("Saving to " + filename + " failed.");
        return 0;
      }
      return 1;
    },

    // save("bla", "file:///name.txt")
    // save("bla", "name.txt")
    // save(tree,  "name.txt") serializes and saves XML (utf-8)
    save: function(tree, url) // HS
    {
      var s;
      if (typeof tree == "string") {
        s = tree;
      }
      else {
        var ser = new XMLSerializer();
        s = ser.serializeToString(tree);
      }
      return this.saveText(s, url);
    }

  };


scripts/computeItemTypeFields.js
function Compute( resultType)
{
  var withCreators = document.getElementById("withCreators").checked;

  var lines = document.getElementById("sql").value;
  // 207	INSERT INTO itemTypes VALUES (1,'note',NULL,0);

  var result;

  // parse itemTypes
  var itRE = /INSERT\s+INTO\s+itemTypes\s+VALUES\s*\(\s*(\d+)\s*,\s*['\"](\w+)[\"']/g;
  var itn2s = {}; // itemType number to string
  var its2n = {}; // itemType string to number
  var its = []; // itemType strings
  while ( (result = itRE.exec(lines)) != null) {
    itn2s[result[1]] = result[2];
    its2n[result[2]] = result[1];
    its.push(result[2]);
  }
  its.sort();

  // parse fields
  var fRE = /INSERT\s+INTO\s+fields\s+VALUES\s*\(\s*(\d+)\s*,\s*['\"](\w+)[\"']/g;
  var fn2s = {}; // field number to string
  while ( (result = fRE.exec(lines)) != null) {
    fn2s[result[1]] = result[2];
  }

  // parse itemTypeFields
  var itfRE = /INSERT\s+INTO\s+itemTypeFields\s+VALUES\s*\(\s*(\d+)\s*,\s*(\d+)/g;
  var itfn2N = {}; // itemType number to field numbers
  while ( (result = itfRE.exec(lines)) != null) {
    if (!itfn2N[result[1]]) {
      itfn2N[result[1]] = [];
    }
    itfn2N[result[1]].push(result[2]);
  }

  if (withCreators) {
    // parse creatorTypes
    var ctRE = /INSERT\s+INTO\s+creatorTypes\s+VALUES\s*\(\s*(\d+)\s*,\s*['\"](\w+)[\"']/g;
    var ctn2s = {}; // creatorType number to string
    var cts2n = {}; // creatorType string to number
    var cts = []; // creatorType strings
    while ( (result = ctRE.exec(lines)) != null) {
      ctn2s[result[1]] = result[2];
      cts2n[result[2]] = result[1];
      cts.push(result[2]);
    }
    cts.sort();
  
    // parse itemTypeCreatorTypes
    var itctRE = /INSERT\s+INTO\s+itemTypeCreatorTypes\s+VALUES\s*\(\s*(\d+)\s*,\s*(\d+)/g;
    var itctn2N = {}; // itemType number to creator type numbers
    while ( (result = itctRE.exec(lines)) != null) {
      if (!itctn2N[result[1]]) {
        itctn2N[result[1]] = [];
      }
      itctn2N[result[1]].push(result[2]);
    }
  }


  var s = "";
  if (resultType == "html") {
    // compute HTML
    s = "<html><body>";
  
    // table of contents
    s += "<p>\n";
    for (var i = 0; i < its.length; ++i) {
      var itemType = its[i];
      s += "<a href='#it_" + itemType + "'>" + itemType + "</a><br/>\n";
    }
    s += "</p>\n";
  
    // each itemType
    for (var j = 0; j < its.length; ++j) {
      var itemType = its[j];
      var itemTypeNumber = its2n[itemType];
      s += "<h1 id='it_" + itemType + "' name='it_" + itemType + "''>" + itemType + "</h1>\n";

      // each field
      var FieldNumbers = itfn2N[itemTypeNumber];
      if (FieldNumbers) {
        var F = [];
        // each itemTypeField (sorted)
        for (var f = 0; f < FieldNumbers.length; ++f) {
          var fieldNumber = FieldNumbers[f];
          var fieldName = fn2s[fieldNumber];
          F.push(fieldName);
        }
        F.sort();
        s += "<p>\n" + F.join("<br/>\n") + "</p>\n";
      }

      if (withCreators) {
        // each creator type
        var CreatorTypeNumbers = itctn2N[itemTypeNumber];
        if (CreatorTypeNumbers) {
          var CT = [];
          // each itemTypeCreatorType (sorted)
          for (var ct = 0; ct < CreatorTypeNumbers.length; ++ct) {
            var creatorTypeNumber = CreatorTypeNumbers[ct];
            var creatorTypeName = ctn2s[creatorTypeNumber];
            CT.push(creatorTypeName);
          }
          CT.sort();
          s += "<p>\n" + CT.join("<br/>\n") + "</p>\n";
        }
      }
    }
  }
  else if (resultType == "testfile") {
    // compute XML
    s = "<?xml version='1.0'?>\n"
      + "<zotero-import>\n";
  
    // each itemType
    for (var j = 0; j < its.length; ++j) {
      var itemType = its[j];

      // special handling for:
      // attachment
      if (itemType == "attachment") { continue; }

      var itemTypeNumber = its2n[itemType];
      s += "<" + itemType + ">\n"; // needs closing tag - do not if (!FieldNumbers) { continue; }

      // note
      if (itemType == "note") { 
        s += "<text>My example note with <strong>strong</strong> text, a list with\n<ul>\n<li>item1</li>\n<li>item2</li>\n</ul>\n</text>\n";
      }

      // each field
      var FieldNumbers = itfn2N[itemTypeNumber];
      if (FieldNumbers) {
        var F = [];
        // each itemTypeField (sorted)
        for (var f = 0; f < FieldNumbers.length; ++f) {
          var fieldNumber = FieldNumbers[f];
          var fieldName = fn2s[fieldNumber];
          F.push(fieldName);
        }
        F.sort();
        for (var k = 0; k < F.length; ++k) {
          var fieldName = F[k];
          s += "<" + fieldName + ">";
          if (fieldName == "title") {
            s += itemType + ": ";
          }
          s += fieldName + "...</" + fieldName + ">\n";
        }
      }

      if (withCreators) {
        // each creator type
        var CreatorTypeNumbers = itctn2N[itemTypeNumber];
        if (CreatorTypeNumbers) {
          var CT = [];
          // each itemTypeCreatorType (sorted)
          for (var ct = 0; ct < CreatorTypeNumbers.length; ++ct) {
            var creatorTypeNumber = CreatorTypeNumbers[ct];
            var creatorTypeName = ctn2s[creatorTypeNumber];
            CT.push(creatorTypeName);
          }
          CT.sort();
          for (var k = 0; k <CT.length; ++k) {
            var creatorTypeName = CT[k];
            s += "<" + creatorTypeName + ">" + creatorTypeName + "...</" + creatorTypeName + ">\n";
          }
        }
      }

      // add two tags as illustration
      var tagBy2 = "Tag " + (j % 2 ? "even" : "odd");
      s += "<tag>" + tagBy2 + "</tag>\n";

      var tagBy3 = (0 == j % 3) ? "A" : (1 == j % 3 ? "B" : "C");
      s += "<tag>Category " + tagBy3 + "</tag>\n";

      s += "</" + itemType + ">\n";
    } // itemTypes
    s += "</zotero-import>\n";
    s = s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/\n/g, "<br/>\n");
  }

  else if (resultType == "rnc") {
    s = ""
      + "# RelaxNG schema for the \"Zotero Xml\" import format.\n"
//TODO Zotero version
      + "# For: Zotero 2.0\n"
      + "# Helmut Steeb 2010-08-27 hs2010\@bible2.net (insert current year)\n"
      + "# Generator: computeItemTypeFields.js\n"
      + "#\n"
      + "# RESTRICTIONS:\n"
      + "# - only <note> is allowed child elements,\n"
      + "#   but the modelling is not correct, you'll get errors there.\n"
      + "# - All other terminal elements accept \"text\", no details are modelled.\n"
      + "\n\n"
      + "zotero-import = element zotero-import {\n(\n";
    // each itemType
    for (var j = 0; j < its.length; ++j) {
      // assumes attachment is not at j==0
      // NOTE: this allows any order of items, even: bill, book, bill, book
      if (its[j] == "attachment") { continue; }
      s += (j == 0 ? " " : "|" ) + " " + its[j] + "\n";
    }
    s += ")+\n}\n\n";

    var AllFields = [];
    var FieldsProcessed = {};
    for (var j = 0; j < its.length; ++j) {
      var itemType = its[j];

      if (itemType == "attachment") { continue; }

      s += itemType + " = element " + itemType + " {\n";
      // note
      if (itemType == "note") { 
        s += "  \\text\n"; // not rnc "text" but our element "text""
      }
      else {
        // each field
        var itemTypeNumber = its2n[itemType];
        var FieldNumbers = itfn2N[itemTypeNumber];
        if (FieldNumbers) {
          var F = [];
          // each itemTypeField (sorted)
          for (var f = 0; f < FieldNumbers.length; ++f) {
            var fieldNumber = FieldNumbers[f];
            var fieldName = fn2s[fieldNumber];
            F.push(fieldName);
          }
          F.sort();
          for (var k = 0; k < F.length; ++k) {
            var fieldName = F[k];
            // NOTE: this allows any order of fields, grouped by identical fieldname, 
            //   but not e.g.: tag, title, tag
            s += (k == 0 ? " " : "&") + " " + fieldName + "?\n";
            if (!FieldsProcessed[fieldName]) {
              AllFields.push(fieldName);
              FieldsProcessed[fieldName] = true;
            }
          }
        }

        if (withCreators) {
          // each creator type
          var CreatorTypeNumbers = itctn2N[itemTypeNumber];
          if (CreatorTypeNumbers) {
            var CT = [];
            // each itemTypeCreatorType (sorted)
            for (var ct = 0; ct < CreatorTypeNumbers.length; ++ct) {
              var creatorTypeNumber = CreatorTypeNumbers[ct];
              var creatorTypeName = ctn2s[creatorTypeNumber];
              CT.push(creatorTypeName);
            }
            CT.sort();
            for (var k = 0; k <CT.length; ++k) {
              var creatorTypeName = CT[k];
              s += "& " + creatorTypeName + "*\n";
              if (!FieldsProcessed[creatorTypeName]) {
                AllFields.push(creatorTypeName);
                FieldsProcessed[creatorTypeName] = true;
              }
            }
          }
        } // withCreators
      } // else each field

      s += "& tag*\n";
      if (itemType != "note") {
        s += "& note*\n";
      }
      s += "& link*\n";
      s += "}\n\n";
    } // itemTypes

    s += "\n";
    s += "tag = element tag { text }\n";
    s += "\\text = element text { (text | ANY)* }\n";
    s += "ANY = element * { ANY }\n";
    s += "link = element link { text }\n";
    s += "\n";

    AllFields.sort();
    for (var f = 0, len = AllFields.length; f < len; ++f) {
      s += AllFields[f] + " = element " + AllFields[f] + " { text }\n";
    }
    s += "\nstart = zotero-import\n";
    s = "<pre>\n" + s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/\n/g, "\n") + "</pre>\n";
  }
  else {
    s = "Unknown result type";      
  }
  var Result = document.getElementById("Result");
  Result.innerHTML = s;
}  
